contracts/allocators/OnsenAllocatorV2.sol

```diff
@@ -0,0 +1,174 @@
+import "../libraries/Address.sol";
+
+// types
+import "../types/BaseAllocator.sol";
+
+struct UserInfo {
+    uint256 amount; // How many LP tokens the user has provided.
+    uint256 rewardDebt;
+}
+
+interface IMasterChef {
+    function pendingSushi(uint256 _pid, address _user) external view returns (uint256);
+
+    function deposit(uint256 _pid, uint256 _amount) external;
+
+    function withdraw(uint256 _pid, uint256 _amount) external;
+
+    function emergencyWithdraw(uint256 _pid) external;
+
+    function userInfo(uint256 _pid, address _user) external view returns (UserInfo memory);
+}
+
+interface ISushiBar {
+    function enter(uint256 _amount) external;
+
+    function leave(uint256 _share) external;
+}
+
+/**
+ *  Contract deploys liquidity from treasury into the Onsen program,
+ *  earning $SUSHI that can be staked and/or deposited into the treasury.
+ */
+contract OnsenAllocatorV2 is BaseAllocator {
+    using SafeERC20 for IERC20;
+
+    /* ========== STATE VARIABLES ========== */
+
+    address immutable sushi; // $SUSHI token
+    address immutable xSushi; // $xSUSHI token
+
+    address immutable masterChef; // Onsen contract
+
+    // address immutable treasury;
+    address public constant treasury = 0x9A315BdF513367C0377FB36545857d12e85813Ef; // Olympus Treasury
+
+    /* ========== CONSTRUCTOR ========== */
+
+    constructor(
+        address _chef,
+        address _sushi,
+        address _xSushi,
+        AllocatorInitData memory data
+    ) BaseAllocator(data) {
+        require(_chef != address(0));
+        masterChef = _chef;
+        require(_sushi != address(0));
+        sushi = _sushi;
+        require(_xSushi != address(0));
+        xSushi = _xSushi;
+
+        // approve for safety, yes toke is being instantly sent to treasury and that is fine
+        // but to be absolutely safe this one approve won't hurt
+        IERC20(sushi).approve(address(extender), type(uint256).max);
+        IERC20(xSushi).approve(address(extender), type(uint256).max);
+    }
+
+    /**
+     1- Deposit LP token into onsen, if deposit is succesfull we should get in return sushi tokens
+     2- Stake the Sushi rewards tokens returned from deposit LP tokens and as rewards form onsen LP stake, we should get in return xSushi tokens
+     3- Calculate gains/loss
+     */
+    function _update(uint256 id) internal override returns (uint128 gain, uint128 loss) {
+        // reads
+        uint256 index = tokenIds[id];
+        IERC20 LPtoken = _tokens[index];
+        uint256 balance = LPtoken.balanceOf(address(this));
+
+        // interactions
+        //Deposit LP token into onsen, if deposit succesfull this address should have in return sushi tokens
+        if (balance > 0) {
+            IERC20(LPtoken).approve(masterChef, balance);
+            IMasterChef(masterChef).deposit(id, balance); // deposit into Onsen
+        }
+
+        // Stake the sushi tokens
+        if (IERC20(sushi).balanceOf(address(this)) > 0) {
```

Comment:
No point of checking this, since the same check happens inside `enterSushiBar`

Reply:
Code updated to avoid the double-check of sushi balance